pkg/expr/func.go

```diff
@@ -0,0 +1,242 @@
+			name:     "NEQ",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.VALUE_TYPE_UNSPECIFIED},
+		},
+		invert: true,
+	}
+}
+
+func (f *eqFunc) Call(args []interface{}) interface{} {
+	res := f.call(args)
+	if f.invert {
+		return !res
+	}
+	return res
+}
+
+func (f *eqFunc) call(args []interface{}) bool {
+	var s0 string
+	var s1 string
+	var ok bool
+
+	if s0, ok = args[0].(string); !ok {
+		return args[0] == args[1]
+	}
+	if s1, ok = args[1].(string); !ok {
+		// s0 is string and s1 is not
+		return false
+	}
+	return matchWithWildcards(s0, s1)
+}
+
+// matchWithWildcards     s0 == ns1.*   --> ns1 should be a prefix of s0
+// s0 == *.ns1 --> ns1 should be a suffix of s0
+// TODO allow escaping *
+func matchWithWildcards(s0 string, s1 string) bool {
+	if strings.HasSuffix(s1, "*") {
+		return strings.HasPrefix(s0, s1[:len(s1)-1])
+	}
+	if strings.HasPrefix(s1, "*") {
+		return strings.HasSuffix(s0, s1[1:])
+	}
+	return s0 == s1
+}
+
+type lAndFunc struct {
+	*baseFunc
+}
+
+// newLANDFunc returns a binary logical AND fn.
+func newLANDFunc() Func {
+	return &lAndFunc{
+		&baseFunc{
+			name:     "LAND",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.BOOL, config.BOOL},
+			nullArgs: false,
+		},
+	}
+}
+
+// Call returns true if both elements true
+// may panic
+func (f *lAndFunc) Call(args []interface{}) interface{} {
+	return args[0].(bool) && args[1].(bool)
+}
+
+type lOrFunc struct {
+	*baseFunc
+}
+
+// newLORFunc returns logical OR fn.
+func newLORFunc() Func {
+	return &lOrFunc{
+		&baseFunc{
+			name:     "LOR",
+			retType:  config.BOOL,
+			argTypes: []config.ValueType{config.BOOL, config.BOOL},
+			nullArgs: false,
+		},
+	}
+}
+
+// Call should return thru if at least one element is true
+func (f *lOrFunc) Call(args []interface{}) interface{} {
+	return args[0].(bool) || args[1].(bool)
+}
+
+// applies to non bools.
+type orFunc struct {
+	*baseFunc
+}
+
+// newORFunc returns an OR fn that selects first non empty argument.
+// types can be anything, but all args must be of the same type.
+func newORFunc() Func {
+	return &orFunc{
+		baseFunc: &baseFunc{
+			name:     "OR",
+			retType:  config.VALUE_TYPE_UNSPECIFIED,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.VALUE_TYPE_UNSPECIFIED},
+			nullArgs: true,
+		},
+	}
+}
+
+// Call selects first non empty argument
+// may return nil
+func (f *orFunc) Call(args []interface{}) interface{} {
+	if args[0] == nil {
+		return args[1]
+	}
+	v := reflect.ValueOf(args[0])
+	if v.Interface() != reflect.Zero(v.Type()).Interface() {
+		return args[0]
+	}
+	return args[1]
+}
+
+// func (Value) MapIndex
+type indexFunc struct {
+	*baseFunc
+}
+
+// newIndexFunc returns a map accessor fn.
+func newIndexFunc() Func {
+	return &indexFunc{
+		baseFunc: &baseFunc{
+			name:     "INDEX",
+			retType:  config.VALUE_TYPE_UNSPECIFIED,
+			argTypes: []config.ValueType{config.VALUE_TYPE_UNSPECIFIED, config.STRING},
```

Comment:
Shouldn't it be used here then? The first arg must be a map, not an unspecified value.